numpy/core/src/multiarray/shape.c

```diff
@@ -126,22 +121,21 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
         ((PyArrayObject_fields *)self)->data = new_data;
     }
 
-    if ((newsize > oldsize) && PyArray_ISWRITEABLE(self)) {
+    if ((newnbytes > oldnbytes) && PyArray_ISWRITEABLE(self)) {
         /* Fill new memory with zeros */
-        elsize = PyArray_DESCR(self)->elsize;
         if (PyDataType_FLAGCHK(PyArray_DESCR(self), NPY_ITEM_REFCOUNT)) {
             PyObject *zero = PyInt_FromLong(0);
             char *optr;
-            optr = PyArray_BYTES(self) + oldsize*elsize;
-            n = newsize - oldsize;
+            optr = PyArray_BYTES(self) + oldnbytes;
+            n = (newnbytes - oldnbytes) / elsize;
```

Comment:
The idea here is that if the number of bytes are different, the element must be non empty. I guess rather than doing the division, we could just calculate this directly though